sympy/matrices/normalforms.py

```diff
@@ -142,3 +142,154 @@ def clear_row(m):
     else:
         result = invs + (m[0,0],)
     return tuple(result)
+
+
+
+def _nonzero(m):
+    """
+    returns a list of containing (i, j) representing the indices of
+    non-zero elements in m.
+    Help function for hermite_norlal_form
+    """
+
+    idx = []
+    for i in range(m.shape[0]):
+        for j in range(m.shape[1]):
+            if m[i, j] != 0:
+                idx.append((i, j))
+    return idx
+
+def _nonzero_negative(m):
+    """
+    returns 0 if the first nonzero column j of A contains more than one nonzero
+    entry, or contains only one nonzero entry and which is positive and  returns 1
+    if the first nonzero column j of A contains only one nonzero entry, which
+    is negative.
+    """
+    columns = list(zip(*_nonzero(m)))[1]
+    nonzero_col = m[:, min(columns)]
+    nums = [i for i in nonzero_col if i !=0]
+    return len(nums) == 1 and nums[0] < 0
+
+
+def hermite_normal_form(m):
+    """
+    Returns the hermite_normal_form of integers matices.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.solvers import RawMatrix as Matrix
+    >>> from sympy.matrices.normalforms import hermite_normal_form
+    >>> m = Matrix([[2, 3, 6, 2], [5, 6, 1, 6], [8, 3, 1, 1]])
+    >>> print(hermite_normal_form(m))
+    (Matrix([
+    [1, 0, 50, -11],
+    [0, 3, 28,  -2],
+    [0, 0, 61, -13]]), Matrix([
+    [ 9, -5, 1],
+    [ 5, -2, 0],
+    [11, -6, 1]]))
+
+    """
+
+    from sympy import eye, ones, Matrix
+
+    row = m.shape[0]
+    col = m.shape[1]
+
+    A = m
+    B = eye(row)
+    L = zeros(row, row)
+    D = ones(row + 1, 1)
+    if row == 1:
+        return m, Matrix([-1]) if _nonzero_negative(m) else eye(1)
+
+    if _nonzero_negative(m):
+        B[row - 1, row - 1] = -1
+        A[row - 1, :] *= -1
+    k = 1
+    while(k < row):
+        col1, col2 = _reduce_matrix(A, B, L, k, k - 1, D)
+        u = (int(D[k - 1]) * int(D[k + 1]) +
+                  int(L[k, k - 1]) * int(L[k, k - 1]))
+        v = int(D[k]) * int(D[k])
```

Comment:
I'd also comment about type casting in this.

Reply:
Done.